src/lib/mina_base/parties_logic.ml

```diff
@@ -0,0 +1,303 @@
+        'bool * 'ledger * 'account * 'inclusion_proof
+        -> ( 'ledger
+           , < bool: 'bool
+             ; ledger: 'ledger
+             ; inclusion_proof: 'inclusion_proof
+             ; account: 'account
+             ; .. > )
+           t
+    | Modify_global_excess :
+        'global_state * ('amount -> 'amount)
+        -> ( 'global_state
+           , < global_state: 'global_state ; amount: 'amount ; .. > )
+           t
+    | Modify_global_ledger :
+        'global_state * ('ledger -> 'ledger)
+        -> ( 'global_state
+           , < global_state: 'global_state ; ledger: 'ledger ; .. > )
+           t
+    | Party_token_id :
+        'party
+        -> ('token_id, < party: 'party ; token_id: 'token_id ; .. >) t
+    | Check_auth_and_update_account :
+        { is_start: 'bool
+        ; protocol_state_predicate: 'protocol_state_predicate option
+        ; party: 'party
+        ; account: 'account
+        ; all_parties: 'all_parties
+        ; inclusion_proof: 'ip }
+        -> ( 'account * 'bool
+           , < inclusion_proof: 'ip
+             ; bool: 'bool
+             ; party: 'party
+             ; all_parties: 'all_parties
+             ; account: 'account
+             ; protocol_state_predicate: 'protocol_state_predicate
+             ; .. > )
+           t
+    | Balance :
+        'account
+        -> ('amount, < account: 'account ; amount: 'amount ; .. >) t
+    | Finalize_local_state :
+        'bool * 'local_state
+        -> (unit, < local_state: 'local_state ; bool: 'bool ; .. >) t
+end
+
+type 'e handler = {perform: 'r. ('r, 'e) Eff.t -> 'r}
+
+module type Inputs_intf = sig
+  module Bool : Bool_intf
+
+  module Ledger : Ledger_intf with type bool := Bool.t
+
+  module Account : Iffable with type bool := Bool.t
+
+  module Amount : Amount_intf with type bool := Bool.t
+
+  module Token_id : Token_id_intf with type bool := Bool.t
+
+  module Parties : Parties_intf with type bool := Bool.t
+end
+
+module Make (Inputs : Inputs_intf) = struct
+  open Inputs
+  module Ps = Inputs.Parties
+
+  let apply (type global_state)
+      (* (step_or_start : [`Start of Inputs.Parties.t | `Step]) *)
+      start_parties protocol_state_predicate
+      (h :
+        (< global_state: global_state
+         ; all_parties: Ps.t
+         ; amount: Amount.t
+         ; bool: Bool.t
+         ; .. >
+         as
+         'env)
+        handler)
+      ((global_state : global_state), (local_state : _ Local_state.t)) =
+    let open Inputs in
+    let is_start = Ps.is_empty local_state.parties in
+    let (party, remaining), local_state =
+      let to_pop =
+        Ps.if_ is_start ~then_:start_parties ~else_:local_state.parties
+      in
+      let party, remaining = Ps.pop to_pop in
+      let local_state =
+        { local_state with
+          all_parties=
+            Ps.if_ is_start ~then_:remaining ~else_:local_state.parties
+        ; token_id=
+            Token_id.if_ is_start ~then_:Token_id.default
```

Comment:
It may not be since `local_state.token_id` will be whatever it was at the end of the last transaction. The idea was that we should clear that out and set it to default when we start a new transaction. 

I think there should be an assertion that the token of the "start" party is default though, like
```ocaml
Bool.(assert_ (not is_start || Token_id.equal party_token Token_id.default)
```